frontend/gateway/gateway.go

```diff
@@ -310,29 +349,51 @@ func (lbf *llbBridgeForwarder) Solve(ctx context.Context, req *pb.SolveRequest)
 		return nil, err
 	}
 
-	exp := map[string][]byte{}
-	if err := json.Unmarshal(req.ExporterAttr, &exp); err != nil {
-		return nil, err
-	}
+	ids := make(map[string]string)
 
-	if expResp != nil {
-		for k, v := range expResp {
-			exp[k] = v
+	lbf.mu.Lock()
+	for k, ref := range res {
+		id := identity.NewID()
+		if ref == nil {
+			id = ""
 		}
+		lbf.refs[id] = ref
+		ids[k] = id
 	}
-
-	id := identity.NewID()
-	lbf.mu.Lock()
-	lbf.refs[id] = ref
 	lbf.mu.Unlock()
+
+	if _, ok := res["default"]; !ok && len(res) != 0 && !req.AllowMapReturn {
+		// this should never happen because old client shouldn't make a map request
+		return nil, errors.Errorf("solve did not return default result")
+	}
+
+	exp := map[string][]byte{}
+	// compatibility mode for older clients
 	if req.Final {
-		lbf.lastRef = ref
+		if err := json.Unmarshal(req.ExporterAttr, &exp); err != nil {
+			return nil, err
+		}
+
+		if expResp != nil {
```

Comment:
I thought this was not needed, if expResp is nil, range expResp will not loop nor panic.